hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java

```diff
@@ -151,208 +149,238 @@ public PreHandleContext status(@NonNull final ResponseCodeEnum status) {
+        if (contractID == null) {
+            return this;
+        }
 
-    /**
-     * Checks if the metadata is already failed. In this case no need to look up that contract's key If the payer key
-     * has not been set we don't add other keys.
-     *
-     * @param id given contract
-     * @return true if the lookup is not needed, false otherwise
-     */
-    private boolean isNotNeeded(@NonNull final ContractID id) {
-        return id.equals(ContractID.DEFAULT) || designatesContractRemoval(id) || status != OK || payerKey == null;
-    }
+        // If an accountID is specified, then the account MUST exist
+        final var account = accountAccess.getContractById(contractID);
+        if (account == null) {
+            throw new PreCheckException(responseCode);
+        }
 
-    /**
-     * Checks if the accountId is a sentinel id 0.0.0
-     *
-     * @param id given accountId
-     * @return true if the given accountID is
-     */
-    private boolean designatesAccountRemoval(@NonNull final AccountID id) {
-        return id.shardNum() == 0 && id.realmNum() == 0 && id.accountNumOrElse(0L) == 0 && !id.hasAlias();
+        // If the account exists but does not require a signature, then there is no key to require.
+        if (!account.isReceiverSigRequired()) {
+            return this;
+        }
+
+        // We will require the key. If the key isn't present, then we will throw the given response code.
+        final var key = account.getKey();
+        if (key == null
+                || key.isEmpty()) { // Or if it is a Contract Key? Or if it is an empty key? Or a KeyList with no
+            // keys? Or KeyList with Contract keys only?
+            throw new PreCheckException(responseCode);
+        }
+
+        return requireKey(key);
     }
 
     /**
-     * Checks if the contractId is a sentinel id <code>0.0.0</code>
+     * Creates a new {@link PreHandleContext} for a nested transaction. The nested transaction will be set on
+     * this context as the "inner context". There can only be one such at a time. The inner context is returned
+     * for convenience.
      *
-     * @param id given contractId
-     * @return true if the given contractId is
+     * @param nestedTxn the nested transaction
+     * @param payerForNested the payer for the nested transaction
+     * @param responseCode the response code to be used if a {@link PreCheckException} is thrown
+     * @return the inner context
+     * @throws PreCheckException If the payer is not valid
      */
-    private boolean designatesContractRemoval(@NonNull final ContractID id) {
-        return id.shardNum() == 0 && id.realmNum() == 0 && id.contractNumOrElse(0L) == 0 && !id.hasEvmAddress();
+    @NonNull
+    public PreHandleContext createNestedContext(
+            @NonNull final TransactionBody nestedTxn,
+            @NonNull final AccountID payerForNested,
+            @NonNull final ResponseCodeEnum responseCode)
```

Comment:
This is a little odd.  It probably needs some clarifying documentation in the javadoc.  Why would I throw a different response code for an invalid payer account in this case?  I'm sure there is some code that uses it, but the javadoc should explain at least that the response code is used if the payer account is invalid or lacks a key, rather than that it's generically "used if an exception is thrown".
Basically, a caller doesn't know how to use this method correctly without reading a few levels down in code, and that's going to create confusion and incorrect usage over time.
If it's possible to elide the response code parameter, so much the better, as it's one of those things that invites incorrect usage.

Reply:
I agree this method is sketchy. It is only used by scheduled service, so on your scheduled service work you might end up with better insight over how this idea of nested contexts should work. Or if it can be eliminated entiirely.